File: experiments/toric_inversions_sample.c
/*
 * Structured + random sampling of I(pi) for n beyond the exhaustive range
 * of toric_inversions_scan.c (H13-I / C33). Uses the same O(n^2) gradient
 * identity (C37) per permutation, so n up to a few hundred is cheap.
 *
 * Families sampled per n: all n reflections pi(i) = (h - i) mod n; affine
 * maps pi(i) = (a*i + b) mod n for a coprime with n; uniform random
 * permutations; reflections perturbed by 1-3 random transpositions.
 *
 * Goal: look for any pi with I(pi) > floor((n-1)^2/4) at n where exhaustive
 * search is infeasible (n >= 13), per AGENTS.md rule 9 (sampling ladder for
 * large n). This is a search for counterexamples, not a proof.
 *
 * Usage: ./toric_inversions_sample [n1 n2 ...]  (defaults to a built-in list)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static int inv_count(const int *a, int n) {
    long long c = 0;
    for (int i = 0; i < n; i++)
        for (int j = i + 1; j < n; j++)
            if (a[i] > a[j]) c++;
    return (int)c;
}

/* O(n^2) via the C37 gradient identity; O(n) extra memory. */
static long long compute_I(const int *pi, int n, const int *invpi) {
    long long *row = malloc(sizeof(long long) * n);
    long long g00 = inv_count(pi, n);
    long long mn = g00;
    long long ga0 = g00;
    for (int a = 0; a < n; a++) {
        row[0] = ga0;
        if (row[0] < mn) mn = row[0];
        for (int b = 0; b < n - 1; b++) {
            int jstar = invpi[b] - a;
            jstar %= n; if (jstar < 0) jstar += n;
            row[b + 1] = row[b] + (n - 1 - 2 * jstar);
            if (row[b + 1] < mn) mn = row[b + 1];
        }
        if (a < n - 1) ga0 += n - 1 - 2 * pi[a];
    }
    free(row);
    return mn;
}

static void fisher_yates(int *a, int n) {
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int t = a[i]; a[i] = a[j]; a[j] = t;
    }
}

static void scan_n(int n, int random_trials) {
    int *pi_ = malloc(sizeof(int) * n);
    int *invpi_ = malloc(sizeof(int) * n);
    long long bound = (long long)(n - 1) * (n - 1) / 4;
    long long worst = -1;
    int worst_kind = -1; /* 0=reflection 1=affine 2=random 3=perturbed reflection */

    for (int h = 0; h < n; h++) {
        for (int i = 0; i < n; i++) { int v = ((h - i) % n + n) % n; pi_[i] = v; }
        for (int i = 0; i < n; i++) invpi_[pi_[i]] = i;
        long long val = compute_I(pi_, n, invpi_);
        if (val > worst) { worst = val; worst_kind = 0; }
    }

    for (int a = 2; a < n && a < 60; a++) {
        int step_b = n > 40 ? (n / 20) + 1 : 1;
        for (int b = 0; b < n; b += step_b) {
            int *chk = calloc(n, sizeof(int));
            int ok = 1;
            for (int i = 0; i < n; i++) {
                int v = ((long long)a * i + b) % n; if (v < 0) v += n;
                pi_[i] = v;
                if (chk[v]) { ok = 0; }
                chk[v] = 1;
            }
            free(chk);
            if (!ok) continue;
            for (int i = 0; i < n; i++) invpi_[pi_[i]] = i;
            long long val = compute_I(pi_, n, invpi_);
            if (val > worst) { worst = val; worst_kind = 1; }
        }
    }

    for (int t = 0; t < random_trials; t++) {
        for (int i = 0; i < n; i++) pi_[i] = i;
        fisher_yates(pi_, n);
        for (int i = 0; i < n; i++) invpi_[pi_[i]] = i;
        long long val = compute_I(pi_, n, invpi_);
        if (val > worst) { worst = val; worst_kind = 2; }
    }

    for (int t = 0; t < 500; t++) {
        int h = rand() % n;
        for (int i = 0; i < n; i++) pi_[i] = ((h - i) % n + n) % n;
        int k = 1 + rand() % 3;
        for (int s = 0; s < k; s++) {
            int x = rand() % n, y = rand() % n;
            int tmp = pi_[x]; pi_[x] = pi_[y]; pi_[y] = tmp;
        }
        for (int i = 0; i < n; i++) invpi_[pi_[i]] = i;
        long long val = compute_I(pi_, n, invpi_);
        if (val > worst) { worst = val; worst_kind = 3; }
    }

    printf("n=%d bound=%lld worst=%lld kind=%d %s\n",
           n, bound, worst, worst_kind, worst <= bound ? "OK" : "FAIL");
    free(pi_); free(invpi_);
}

int main(int argc, char **argv) {
    srand(12345);
    if (argc > 1) {
        for (int i = 1; i < argc; i++) scan_n(atoi(argv[i]), 3000);
        return 0;
    }
    int ns[] = {13, 14, 15, 16, 18, 20, 24, 30, 40, 50, 64, 80, 100, 101, 127, 150, 200};
    for (unsigned i = 0; i < sizeof(ns) / sizeof(ns[0]); i++) scan_n(ns[i], 3000);
    return 0;
}

File: experiments/toric_inversions_sample.py
"""toric_inversions_sample.py — structured/random samples of I(pi) at large n
(AGENTS.md rule 9 ladder step: n = 20, 50, 100, 101 and neighbourhoods).

I(pi) = min over double cuts (q, c) of the number of inversions of the
relabelled line (definition: docs/notes/h13_line_model.md §0). Computed with
the same O(n^2) row/column recursion as experiments/toric_inversions_fast.c
(SAMPLED here, not exhaustive: this checks specific pi, not all of S_n).

Families: both reflections (h = 0, 1), the antipodal transposition (n even),
invertible affine maps pi(i) = a*i+b mod n, uniformly random permutations
(fixed seed), identity, rev_n. This is error-finding, not a proof (rule 9).

Usage: python3 experiments/toric_inversions_sample.py
Version toric_inversions_sample-1.0.
"""

import math
import random


def I_fast(pi):
    n = len(pi)
    pinv = [0] * n
    for i, v in enumerate(pi):
        pinv[v] = i

    def inv0(seq):
        m = len(seq)
        c = 0
        for i in range(m):
            for j in range(i + 1, m):
                if seq[i] > seq[j]:
                    c += 1
        return c

    base = inv0(pi)
    row0 = [0] * n
    row0[0] = base
    for b in range(n - 1):
        r = pinv[b]
        row0[b + 1] = row0[b] + (n - 1) - 2 * r
    best = 10 ** 9
    for b in range(n):
        cur = row0[b]
        if cur < best:
            best = cur
        for a in range(n - 1):
            r = pi[a] - b
            if r < 0:
                r += n
            cur = cur + (n - 1) - 2 * r
            if cur < best:
                best = cur
    return best


def check(name, pi, n, rows):
    bound = ((n - 1) ** 2) // 4
    val = I_fast(pi)
    status = "OK" if val <= bound else "VIOLATION"
    rows.append((n, name, val, bound, status))
    print(f"n={n:4d} {name:30s} I={val:6d} bound={bound:6d} {status}")


def main():
    rows = []
    for n in (20, 50, 100, 101):
        check("reflection h=0", [(-i) % n for i in range(n)], n, rows)
        check("reflection h=1", [(1 - i) % n for i in range(n)], n, rows)
        if n % 2 == 0:
            pi = list(range(n))
            m = n // 2
            pi[m], pi[m + 1] = pi[m + 1], pi[m]
            check("antipodal transposition", pi, n, rows)
        for a, b in [(2, 0), (3, 1), (5, 7), (n - 2, 3)]:
            if math.gcd(a, n) != 1:
                continue
            check(f"affine a={a},b={b}", [(a * i + b) % n for i in range(n)], n, rows)
        random.seed(1)
        for trial in range(3):
            pi = list(range(n))
            random.shuffle(pi)
            check(f"random#{trial}", pi, n, rows)
        check("identity", list(range(n)), n, rows)
        check("rev_n", [n - 1 - i for i in range(n)], n, rows)
    violations = [r for r in rows if r[4] == "VIOLATION"]
    print(f"\ntotal checks: {len(rows)}, violations: {len(violations)}")


if __name__ == "__main__":
    main()
